### ISSUE TITLE
ASAN and valgrind detect memory leaks in our executables

### ISSUE THREAD

**Issue by nikwit** [sxs-collaboration/spectre#3492]
# Bug reports:

Both ASAN and valgrind detect memory leaks in our executables. ASAN traces these as follows:

 - starting `in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)`
 -  ending `in start_thread /build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:477`
 - and going mostly through `Algorithm` and `DataBox` related headers in the middle

It seems that charm spawns pthreads and does not join them before terminating the program which makes ASAN/valgrind think there is a memory leak ( https://stackoverflow.com/questions/5610677/valgrind-memory-leak-errors-when-using-pthread-create )

This is a bit annoying because it makes it tough to actually detect memory leaks in our code. Also, it does not appear to trigger our CI test with ASAN even though the exit code of running an executable with a detected memory leak is 1 on my system.

**Comment by nilsvu** [comment]
I found this snippet in `Tests.yaml:526`:

```yaml
          # We disable ASAN's leak sanitizer because Charm++ has false
          # positives that would cause the build to fail. We disable
          # leak sanitizer for the ctest runs inside CMake anyway.
```

Indeed, cmake sets `ASAN_OPTIONS=detect_leaks=0` in a bunch of places. That's presumably why CI doesn't fail. @nilsdeppe do you have some insight here, since I think you added the snippet?

**Comment by nilsdeppe** [comment]
yep, Charm++ and OpenMPI have oodles of memory leaks (things that are cleaned up when the process closes and don't waste a lot of memory, though) and so we basically can't do leak checking. This is an upstream issue.
